src/io/delimited.cr

```diff
@@ -46,6 +46,147 @@ class IO::Delimited < IO
     check_open
     return 0 if @finished
 
+    read_internal(slice)
+  end
+
+  private def read_internal(slice : Bytes) : Int32
+    if peek = @io.peek
+      read_with_peek(slice, peek)
+    else
+      read_without_peek(slice)
+    end
+  end
+
+  private def read_with_peek(slice : Bytes, peek : Bytes) : Int32
+    # If there's nothing else to peek, we reached EOF
+    if peek.empty?
+      @finished = true
+
+      if @active_delimiter_buffer.empty?
+        return 0
+      else
+        # If we have something in the active delimiter buffer,
+        # but we don't have any more data to read, that wasn't
+        # the delimiter so we must include it in the slice.
+        slice.copy_from(@active_delimiter_buffer)
+        return @active_delimiter_buffer.size
+      end
+    end
+
+    # If we have something in the active delimiter buffer
+    unless @active_delimiter_buffer.empty?
+      # This is the rest of the delimiter we have to match
+      delimiter_remaining = @read_delimiter[@active_delimiter_buffer.size..]
+
+      # This is how much we can actually match of that (peek might not have enough data!)
+      min_size = Math.min(delimiter_remaining.size, peek.size)
+
+      # See if what remains to match in the delimiter matches whatever
+      # we have in peek, limited to what's available
+      if delimiter_remaining[0, min_size] == peek[0, min_size]
+        # If peek has enough data to match the entire rest of the delimiter...
+        if peek.size >= delimiter_remaining.size
+          # We found the delimiter!
+          @io.skip(min_size)
+          @finished = true
+          return 0
+        else
+          # Copy the remaining of peek to the active delimiter buffer for now
+          (@delimiter_buffer + @active_delimiter_buffer.size).copy_from(peek)
+          @active_delimiter_buffer = @delimiter_buffer[0, @active_delimiter_buffer.size + peek.size]
+
+          # Skip whatever we had in peek, and try reading more
+          @io.skip(peek.size)
+          return read_internal(slice)
+        end
+      else
+        # No match: just copy from the active delimiter
+        read_bytes = @active_delimiter_buffer.size
+        slice.copy_from(@active_delimiter_buffer)
+        slice += read_bytes
+        @active_delimiter_buffer = Bytes.empty
+        return read_bytes + read_internal(slice)
```

Comment:
I'm wondering about the case that the active delimiter buffer contains part of the delimiter, but doesn't start with it. This can happen when the delimiter contains a duplicate sequence.

For example, let's assume the delimiter is `aab`. `@active_delimiter_buffer` contains `aa`. Now, `peek` starts with `ab`. That means, the next character does not match the end of the delimiter. But we should still detect the delimiter starting at the second position of `@active_delimiter_buffer`.
```
                    delimiter
                      ---
                     aaab 
                     --~~
active_delimter_buffer peek
```

It looks like this is handled correctly in the non-peek implementation, but I haven't tested it yet.

Reply:
Done!